File: frontend/types/lesson.ts
export type ExerciseType =
  | 'multiple_choice'
  | 'translate_to_target'
  | 'translate_to_source'
  | 'match_pairs'
  | 'fill_in_blank'
  | 'type_answer';

export interface MultipleChoiceOption {
  id: string;
  text: string;
  image?: string;
}

export interface MatchPair {
  left: string;
  right: string;
}

export interface ExerciseContent {
  options?: MultipleChoiceOption[];
  source_sentence?: string;
  target_sentence?: string;
  word_bank?: string[];
  pairs?: MatchPair[];
  sentence_parts?: string[];
  blank_options?: string[];
  hint?: string;
}

export interface Exercise {
  id: number;
  lesson_id: number;
  type: ExerciseType;
  prompt: string;
  content: ExerciseContent;
  order_index: number;
}

export interface StartLessonResponse {
  attempt_id: number;
  lesson_id: number;
  lesson_title: string;
  hearts_remaining: number;
  xp_reward: number;
  exercises: Exercise[];
}

export interface SubmitAnswerRequest {
  exercise_id?: number;
  user_answer: string | string[] | Record<string, string>;
  attempt_id?: number;
}

export interface SubmitAnswerResponse {
  is_correct: boolean;
  correct_answer: string | string[] | Record<string, string>;
  explanation: string | null;
  hearts_remaining: number;
  is_failed: boolean;
  mistakes_count: number;
}

export interface CompleteLessonResponse {
  status: string;
  xp_earned: number;
  total_xp: number;
  current_hearts: number;
  streak_count: number;
  daily_goal_xp: number;
  today_xp: number;
  daily_goal_reached: boolean;
  unlocked_achievements: string[];
  next_lesson_id: number | null;
}

File: backend/app/schemas/lesson.py
from typing import List, Optional, Any
from pydantic import BaseModel, ConfigDict


class ExerciseResponse(BaseModel):
    id: int
    lesson_id: int
    type: str  # multiple_choice, translate_to_target, translate_to_source, match_pairs, fill_in_blank, type_answer
    prompt: str
    content: Any  # JSON object containing options, word tokens, pairs, hints
    order_index: int

    model_config = ConfigDict(from_attributes=True)


class LessonDetailResponse(BaseModel):
    id: int
    skill_id: int
    title: str
    order_index: int
    xp_reward: int
    exercises: List[ExerciseResponse]

    model_config = ConfigDict(from_attributes=True)


class StartLessonResponse(BaseModel):
    attempt_id: int
    lesson_id: int
    lesson_title: str
    hearts_remaining: int
    xp_reward: int
    exercises: List[ExerciseResponse]


class SubmitAnswerRequest(BaseModel):
    exercise_id: Optional[int] = None
    user_answer: Any  # string, array of strings, or dictionary of pairs
    attempt_id: Optional[int] = None


class SubmitAnswerResponse(BaseModel):
    is_correct: bool
    correct_answer: Any
    explanation: Optional[str] = None
    hearts_remaining: int
    is_failed: bool
    mistakes_count: int


class CompleteLessonRequest(BaseModel):
    attempt_id: Optional[int] = None


class CompleteLessonResponse(BaseModel):
    status: str  # "completed"
    xp_earned: int
    total_xp: int
    current_hearts: int
    streak_count: int
    daily_goal_xp: int
    today_xp: int
    daily_goal_reached: bool
    unlocked_achievements: List[str]
    next_lesson_id: Optional[int] = None
